File: bench/parse_float/parse_float.java
// Java peer of bench/parse_float.vader. `Double.parseDouble` is correctly
// rounded, and `doubleToLongBits` gives the pattern the checksum sums.

static final int INPUT_COUNT = 4096;
static final int PASS_COUNT = 160;

static final String[] FRACTIONS = { "5", "25", "125", "0625", "1", "7", "142857", "33333333" };
static final String[] EXPONENTS = { "", "e3", "e-3", "e17", "e-17" };

static String[] buildInputs() {
    String[] out = new String[INPUT_COUNT];
    for (int index = 0; index < INPUT_COUNT; index++) {
        long mantissa = (long) index * 1103515245L % 1000003L;
        out[index] = mantissa + "." + FRACTIONS[index % 8] + EXPONENTS[index % 5];
    }
    return out;
}

void main() {
    String[] inputs = buildInputs();
    long high = 0, low = 0;
    for (int pass = 0; pass < PASS_COUNT; pass++) {
        for (String text : inputs) {
            double parsed = Double.parseDouble(text);
            long bits = Double.doubleToLongBits(parsed);
            high += bits >>> 32;
            low += bits & 0xFFFFFFFFL;
        }
    }
    IO.println("parse_float inputs=%d passes=%d high=%d low=%d"
        .formatted(INPUT_COUNT, PASS_COUNT, high, low));
}

File: bench/parse_float/parse_float.go
// Go peer of bench/parse_float.vader. `strconv.ParseFloat` is correctly rounded,
// which is the property the checksum pins.

package main

import (
	"fmt"
	"math"
	"strconv"
)

const (
	inputCount = 4096
	passCount  = 160
)

var fractions = []string{"5", "25", "125", "0625", "1", "7", "142857", "33333333"}
var exponents = []string{"", "e3", "e-3", "e17", "e-17"}

func buildInputs() []string {
	out := make([]string, 0, inputCount)
	for index := 0; index < inputCount; index++ {
		mantissa := int64(index) * 1103515245 % 1000003
		out = append(out, fmt.Sprintf("%d.%s%s", mantissa, fractions[index%8], exponents[index%5]))
	}
	return out
}

func main() {
	inputs := buildInputs()
	var high, low int64
	for pass := 0; pass < passCount; pass++ {
		for _, text := range inputs {
			parsed, err := strconv.ParseFloat(text, 64)
			if err == nil {
				bits := math.Float64bits(parsed)
				high += int64(bits >> 32)
				low += int64(bits & 0xFFFFFFFF)
			}
		}
	}
	fmt.Printf("parse_float inputs=%d passes=%d high=%d low=%d\n", inputCount, passCount, high, low)
}
